bpf/cilium-probe-kernel-hz.c

```diff
@@ -0,0 +1,180 @@
+// SPDX-License-Identifier: GPL-2.0
+/* Copyright (C) 2020 Authors of Cilium */
+
+#define _GNU_SOURCE
+
+#include <stdio.h>
+#include <unistd.h>
+#include <sched.h>
+#include <errno.h>
+#include <string.h>
+#include <time.h>
+#include <stdint.h>
+#include <stdlib.h>
+
+#include <sys/resource.h>
+
+struct cpu_jiffies {
+	uint64_t *jiffies;
+	uint32_t cpus;
+};
+
+static const uint64_t kernel_hz[] = { 100, 250, 300, 1000 };
+
+#define abs(x)	({ x < 0 ? -x : x; })
+
+static int pin_to_cpu(int cpu)
+{
+	cpu_set_t set;
+
+	CPU_ZERO(&set);
+	CPU_SET(cpu, &set);
+
+	return sched_setaffinity(getpid(), sizeof(set), &set);
+}
+
+static int fix_priority(void)
+{
+	struct sched_param sp = {
+		.sched_priority = sched_get_priority_max(SCHED_FIFO),
+	};
+
+	return sched_setscheduler(0, SCHED_FIFO, &sp);
+}
+
+static FILE *timer_list_open(void)
+{
+	return fopen("/proc/timer_list", "r");
+}
+
+static void timer_list_close(FILE *fp)
+{
+	fclose(fp);
+}
+
+static int prep_kern_jiffies(struct cpu_jiffies *before,
+			     struct cpu_jiffies *after)
+{
+	uint64_t jiffies = 0;
+	char buff[512];
+	int cpus = 0;
+	FILE *fp;
+
+	fp = timer_list_open();
+	while (fp && fgets(buff, sizeof(buff), fp)) {
+		if (sscanf(buff, "jiffies: %lu\n", &jiffies) == 1)
+			cpus++;
+	}
+	timer_list_close(fp);
+
+	if (!cpus) {
+		fprintf(stderr, "No procfs support?\n");
+		return -EIO;
+	}
+
+	before->cpus = after->cpus = cpus;
+	before->jiffies = calloc(cpus, sizeof(*before->jiffies));
+	after->jiffies  = calloc(cpus, sizeof(*before->jiffies));
+
+	if (!before->jiffies || !after->jiffies) {
+		free(before->jiffies);
+		free(after->jiffies);
+		fprintf(stderr, "Error allocating per CPU jiffies: %s\n",
+			strerror(errno));
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int fetch_kern_jiffies(const struct cpu_jiffies *curr)
+{
+	char buff[512];
+	int cpus = 0;
+	FILE *fp;
+
+	fp = timer_list_open();
+	while (fp && fgets(buff, sizeof(buff), fp) && cpus < curr->cpus) {
+		if (sscanf(buff, "jiffies: %lu\n", &curr->jiffies[cpus]) == 1)
+			cpus++;
+	}
+	timer_list_close(fp);
+
+	if (cpus != curr->cpus) {
+		fprintf(stderr, "CPU mismatch when fetching jiffies\n");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int dump_kern_jiffies(const struct cpu_jiffies *fixed,
+			     const struct cpu_jiffies *result)
+{
+	uint64_t delta, warp = 0;
+	int i, j, ret = 0;
+
+	for (i = 0; i < result->cpus; i++) {
+		result->jiffies[i] -= fixed->jiffies[i];
+		for (j = 0, delta = ~0;
+		     j < sizeof(kernel_hz)/sizeof(kernel_hz[0]); j++) {
+			int64_t x = abs((int64_t)(kernel_hz[j] -
+						  result->jiffies[i]));
+			if (x < delta) {
+				delta = x;
+				fixed->jiffies[i] = kernel_hz[j];
+			}
+		}
+	}
+
+	for (i = 0; i < result->cpus; i++) {
+		int64_t x = abs((int64_t)(fixed->jiffies[i] -
+					  result->jiffies[i]));
```

Comment:
Nit: I think the two loops can be merged and you can use `delta` instead of computing `x` a second time.

Reply:
Ah, true, good point. Will fix. :)